simplified framework options exposed to user and added aggregate-like initialization.

pocdn opened the pull request:
Exposing to the user the Instance and Device at Framework level caused some examples to stop working.

  // Initialize makers
  vku::InstanceMaker im{};
  im.defaultLayers();
  vku::DeviceMaker dm{};
  dm.defaultLayers();

The fix is to instead expose Framework high-level options only

  // Define framework options
  vku::FrameworkOptions fo = {
    .useCompute = false;
    .useTessellationShader = false;
    .useGeometryShader = false;
    .useMultiView = false;
  };

and not expose "im" and and "dm" to the user directly. The reason is that all the possible options of those two require commensurate code changes beyond just specifying their member variables and more vk::code changes inside vku_framework.hpp. 

The newer compilers didn't flag correctly error of using aggregate initilizers when a constructor is user defined. However, the consequence affected the uses of  

vku::Window
vk::Viewport
vk::RenderPassBeginInfo
vk::CommandPoolCreateInfo  

The fix was to create thin "Maker" wrappers around all these except vku::Window. There wasn't really much benefit for vku::Window. For the other three the code was returned to a more aggregate-like use case with benefit of allowing setting of internal variables in any order unlike actual c++20 aggregate initializers.

prior:

    displacementRpbi = vk::RenderPassBeginInfo {
      *displacementRenderPass,
      *displacementFrameBuffer,
      vk::Rect2D{{0, 0}, {window.width(), window.height()}},
      (uint32_t) clearColours.size(),
      clearColours.data()
    };

Note, the above can still be used if user likes but now the more aggregate-like use case is also possible (and used in all examples):

    displacementRpbi = vku::RenderPassBeginInfoMaker{}
      .renderPass( *displacementRenderPass )
      .framebuffer( *displacementFrameBuffer )
      .renderArea( vk::Rect2D{{0, 0}, {window.width(), window.height()}} )
      .clearValueCount( (uint32_t) clearColours.size() )
      .pClearValues( clearColours.data() )
      .createUnique();

And an example/swirltexture was added.

Two minor changes related to external libraries which give annoying warnings that could be easily fixed.

external/gilgamesh/mesh.hpp ... simply add a missing return statements, added this proposal to original gilgamesh project.

external/glm/detail/type_half.inl ... latest glm actually does this change from "f *= f" to "f = f * f"

Commit:
added multiview,tesselation&geometry shader capability;multiple examples

Commit:
returned line endings back to dos from unix

Commit:
added swirltexture example. added simplified vku::FrameworkOptions. modified vku.hpp with helper classes to be similar to designated initializers

lhog commented:
Looks cool, but something has happened and lots of files show conflicts.

Commit:
Resolved merge conflict

pocdn commented:
My local branch had become out of sync with this master branch. The conflicts should be gone now.

Commit:
fixed line endings from unix to dos

Commit:
prior end of lines fix failed, now should be fixed

Commit:
added subpass example

Commit:
fixed semaphore and fnece syncronization issue in draw()

Commit:
fixed validation error related to image aquire semaphore

Commit:
fixed access syncronization issue

Commit:
added windowoptions struct

lhog commented:
@andy-thomason would you please grant commit rights to @pocdn 
Seems like none of us has time to contribute to Vookoo :(

andy-thomason commented:
I've added @pocdn as a collablorator. If there is anything else that is required. Please say so. If this was an organisation, I could grant admin rights, but I'm not sure how to do that in a user repo.

We could create an organisation and transfer the repo there if that helps.

andy-thomason commented:
I'm overwhelmed that you all have found this useful. The Vulkan API does not make it easy to get started, which was the primary motive for doing this.

Commit:
trivial std::cout formatting changes

Commit:
added ability to use specialization constants in compute shader. Demonstrated in new example computeDisplay

Commit:
added computeDisplay Example

Commit:
added example showing compute stage writing to image then displayed in graphics pipeline

Commit:
fixed validation error in computeImageDisplay example due to missing memory barrier

Commit:
Removed more validation errors due to syncronization.

Commit:
fixed bug due to out values not _ALL_ being set

Commit:
explicitly upcast iFrame to float(iFrame)

Commit:
* examples/subpass/subpass.cpp

 Replaced two hardcoded 3-element arrays (myFrameBufferPass and myRpbi) with std::vector sized dynamically to swapchainImageCount, fixing a
   crash when the swapchain has fewer than 3 images.

* include/vku/vku_framework.hpp

 Fixed the draw() method to use imageIndex (acquired swapchain slot) instead of currentFrame (CPU ring counter) when indexing
  commandBufferFences_, framebuffers_, staticDrawBuffers_, and the dynamic lambda, preventing rendering to the wrong swapchain image when
  the two counters diverge

* include/vku/vku_framework_sdl2.hpp

 Add vku_framework_sdl2.hpp: GLFW-free variant of vku_framework.hpp for SDL2 builds

  Removes the GLFW Window constructor and GLFW/native headers while retaining
  the non-GLFW Window constructor and the unchanged Framework class.

Commit:
fixed remaining vulkan validation hazards

Commit:
added first SDL2 example

Commit:
need SDL2::SDL2main for windows

Commit:
windows also needs WIN32

Commit:
better fix for SDL main

Commit:
Fixed vulkan.hpp compatibility with newer Vulkan SDKs by replacing three deprecated symbols

Commit:
fixed VK_HEADER_VERSION specific differences

Commit:
bumped from vulkan 1.1 to 1.3 for dynamic rendering. also added fdtd2dUpmlCompute example

Commit:
added two folders to ignore list

Commit:
fixed SYNC-HAZARDs with both new examples

Commit:
fixed SYNC-HAZARD-PRESENT-AFTER-WRTIE seen in windows but not linux

Commit:
more fixes for validation errors

Commit:
added MsaaImage and subpassInputAttachment to vku.h. added SDL2 msaa example.

Commit:
added noisedemo based on github.com/ashima/webgl-noise demo and benchmark

Commit:
added examplesSDL2/fluid3D which needed TextureImage3D added to vku.hpp

Commit:
add ability to cycle thru different inner K loop sizes

Commit:
added PrecomputedAtmosphericScattering examples

Commit:
moved okretnica.ktx to teapot example folder

Commit:
Unified GLFW and SDL2 into a single vku_framework.hpp selected by VKU_GLFW, VKU_SDL2, or VKU_NO_WINDOW, added an SDL_Window* constructor to vku::Window, and updated all examples and project files

Commit:
no need for vku/vku_framework_sdl2.hpp

Commit:
converted dat to ktx files offline to simplify PrecomputedAtmosphericScattering2017.cpp

Commit:
ktx files from dat files

Commit:
commited to ktx file format

Commit:
keeping rw files, although not used

Commit:
added gamedemo to demostrate multiple shaders in practical case

Commit:
slight fix for windows related problem

Commit:
more for windows issue
